### PR TITLE
[Product Collection] Add unique ID attribute

### LINKED ISSUE #42861: Introduce a unique identifier for each Product Collection block
### Description

The aim is to assign a unique identifier to each block in the database, in addition to the `query_id` used within a page or template.
It is important that this new identifier is unique at the database level. This means that it should not be possible for two blocks to end up with the same identifier, whether they are on the same page **or different ones**.

The goal is to allow third-party developers to identify individual block instances **for caching** or other operations that require this unique identification.

### PR DESCRIPTION
**Pull request by imanish003**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

In this update, we introduced a unique identifier for each instance of the Product Collection block. This change involves several key modifications:

1. **Block JSON Update**: Added a new `id` attribute in `block.json` to store the unique identifier for each block instance.

2. **Product Collection Content Update**: In `product-collection-content.tsx`, we utilized the `uuid` library to generate a unique ID. This is done in a `useEffect` hook, ensuring that each block instance receives a distinct ID upon initialization.

3. **Type Definitions**: The `ProductCollectionAttributes` interface in `types.ts` was updated to include the new `id` attribute, aligning with our enhanced data structure.

4. **Package Dependencies**: To support this feature, `@types/uuid` and `uuid` were added to `package.json`, ensuring we have the necessary tools for ID generation.

This enhancement is pivotal for uniquely identifying each Product Collection block, paving the way for more robust and individualized block handling in the future.

Closes #42861 

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add Product Collection block
2. Open the code editor and confirm that the Product Collection block includes an "id" attribute.
3. Add multiple Product Collection blocks and verify that each has a unique “id” attribute.
4. On Frontend, inspect Product Collection block and verify that there is “data-id” attribute in HTML

Here is a quick demo video:

https://github.com/woocommerce/woocommerce/assets/16707866/674e4273-a471-49e0-b8ce-3867d2934048

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 
Add "id" attribute which will be unique for each Product Collection block
</details>

### CODE REVIEW

**Comment by xristos3490** [comment]
After my first pass, I tested the functionality with multiple blocks within the same page and across different pages. It seems to be working great!

Here's how the result looks when adding two collections into a page/template: 

Moving forward, I think this is another case where we need this attribute to be accessible within the block context (both SSR and the editor.) Similar to https://github.com/woocommerce/woocommerce/pull/43247

**Comment by imanish003** [comment]
Hey @xristos3490,

Thank you so much for your review of this draft PR 🙏🏻

> Moving forward, I think this is another case where we need this attribute to be accessible within the block context (both SSR and the editor.) Similar to https://github.com/woocommerce/woocommerce/pull/43247
> 

I'm considering merging this PR and addressing the issue you highlighted in a separate PR. Once we've reached a consensus on a solution here(p1704274893188949-slack-C05UMTA7W0P), we could initiate a dedicated PR for it.

If you're on board with merging this PR as it stands, I'll proceed to mark it as ready for review. Additionally, I plan to include “id” in “providesContext”, which should enable us to access it in SSR for the time being. What do you think?

File: plugins/woocommerce/package.json [hunk 1]
```diff
@@ -222,6 +222,7 @@
 				"node_modules/@woocommerce/e2e-core-tests/CHANGELOG.md",
 				"node_modules/@woocommerce/api/dist/",
 				"node_modules/@woocommerce/admin-e2e-tests/build",
+				"node_modules/@woocommerce/classic-assets/build",
```
Comment (kmanijak) [inline plugins/woocommerce/package.json]: Out of curiosity? What's the purpose of that entry? Was it added automatically? If not, what does it solve?

**Comment by imanish003** [comment]
While merging `trunk` into this branch, I encountered the following error. Consequently, I am reverting this PR to draft status. I will investigate this issue further tomorrow.
image

**Comment by xristos3490** [comment]
> If you're on board with merging this PR as it stands, I'll proceed to mark it as ready for review. Additionally, I plan to include “id” in “providesContext”, which should enable us to access it in SSR for the time being. What do you think?

@imanish003 Sounds good! Handling the editor is a separate challenge altogether!

File: plugins/woocommerce/package.json (hunk 1 above)
Comment (imanish003) [reply plugins/woocommerce/package.json]: This was added automatically when I merged `trunk` into this branch. I am not sure if I should keep this automatically generated file 🤔

Comment (imanish003) [reply plugins/woocommerce/package.json]: I went ahead and removed it in https://github.com/woocommerce/woocommerce/pull/43314/commits/18b7ebf19d7f5e8e3de15879e59bd76da686fa0f as it doesn't make sense to have this change in this PR.

**Comment by imanish003** [comment]
JFYI, here's how I resolved the PNPM error:

1. I first reverted certain files to their original state as in the trunk branch:
    1. `git checkout origin/trunk -- plugins/woocommerce-blocks/package.json`
    2. `git checkout origin/trunk -- plugins/woocommerce/package.json`
    3. `git checkout origin/trunk -- pnpm-lock.yaml`
2. Next, I re-added the `uuid` dependencies:
    1. `pnpm add --filter='@woocommerce/block-library' uuid`
    2. `pnpm add --filter='@woocommerce/block-library' @types/uuid --save-dev`
3. After making these changes, I pushed them.

The PR is now marked as ready for review once again.

CC: @kmanijak

**Review by kmanijak — CHANGES_REQUESTED** [review]
Seems like `pnpm-lock` file was updated with 1.5k lines. I think we should aim to limit the changes to adding a single dependency. Changing versions of so many dependencies may be dangerous and requires much more extensive testing.
image

**Comment by imanish003** [comment]
@kmanijak 
> Seems like `pnpm-lock` file was updated with 1.5k lines. I think we should aim to limit the changes to adding a single dependency. Changing versions of so many dependencies may be dangerous and requires much more extensive testing.

Thanks for raising this concern. It's a point I've been considering and wanted to clarify with #woo-monorepo before merging this PR. Let me clarify this in #woo-monorepo 🤝

**Comment by imanish003** [comment]
Hey @kmanijak,
I managed to minimize the changes in `pnpm-lock.yaml`. Here's what I did:

- Remove `pnpm-lock.yaml` file: `rm pnpm-lock.yaml`
- Regenerate lock file: `pnpm install`

It's a bit odd that I needed to delete and then regenerate the lock file 🤷🏻‍♂️

Could you take another look at the PR when you have a moment? Thanks! 🙇‍♂️

**Review by kmanijak — APPROVED** [review]
Tested the changes and unique ID is added to existing Product Collection blocks as well as the new ones. All seems to be working just fine!

Thanks for your patience in solving the issue with massive `pnpm-lock` file changes! 🚀

**Pull request merged**